donkeycar/parts/interpreter.py

```diff
@@ -0,0 +1,296 @@
+
+        # Get input and output tensors.
+        self.input_details = self.interpreter.get_input_details()
+        self.output_details = self.interpreter.get_output_details()
+
+        # Get Input shape
+        self.input_shapes = []
+        logger.info('Load model with tflite input tensor details:')
+        for detail in self.input_details:
+            logger.debug(detail)
+            self.input_shapes.append(detail['shape'])
+
+    def compile(self, **kwargs):
+        pass
+
+    def invoke(self) -> Sequence[Union[float, np.ndarray]]:
+        self.interpreter.invoke()
+        outputs = []
+        for tensor in self.output_details:
+            output_data = self.interpreter.get_tensor(tensor['index'])
+            # as we invoke the interpreter with a batch size of one we remove
+            # the additional dimension here again
+            outputs.append(output_data[0])
+        # don't return list if output is 1d
+        return outputs if len(outputs) > 1 else outputs[0]
+
+    def predict(self, img_arr, other_arr) \
+            -> Sequence[Union[float, np.ndarray]]:
+        assert self.input_shapes and self.input_details, \
+            "Tflite model not loaded"
+        input_arrays = (img_arr, other_arr)
+        for arr, shape, detail \
+                in zip(input_arrays, self.input_shapes, self.input_details):
+            in_data = arr.reshape(shape).astype(np.float32)
+            self.interpreter.set_tensor(detail['index'], in_data)
+        return self.invoke()
+
+    def predict_from_dict(self, input_dict):
+        for detail in self.input_details:
+            k = detail['name']
+            inp_k = input_dict[k]
+            inp_k_res = inp_k.reshape(detail['shape']).astype(np.float32)
+            self.interpreter.set_tensor(detail['index'], inp_k_res)
+        return self.invoke()
+
+    def get_input_shapes(self):
+        assert self.input_shapes is not None, "Need to load model first"
+        return self.input_shapes
+
+
+class TensorRT(Interpreter):
+    """
+    Uses TensorRT to do the inference.
+    """
+    def __init__(self):
+        self.frozen_func = None
+        self.input_shapes = None
+
+    def get_input_shapes(self) -> List[tf.TensorShape]:
+        return self.input_shapes
+
+    def compile(self, **kwargs):
+        pass
+
+    def load(self, model_path: str) -> None:
+        saved_model_loaded = tf.saved_model.load(model_path,
+                                                 tags=[tag_constants.SERVING])
+        graph_func = saved_model_loaded.signatures[
+            signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY]
+        self.frozen_func = convert_var_to_const(graph_func)
+        self.input_shapes = [inp.shape for inp in graph_func.inputs]
+
+    def predict(self, img_arr: np.ndarray, other_arr: np.ndarray) \
+            -> Sequence[Union[float, np.ndarray]]:
+        # first reshape as usual
+        img_arr = img_arr.reshape((1,) + img_arr.shape).astype(np.float32)
```

Comment:
Use `unsqueeze()` here and above. Its a lot more easier to understand. Also a comment would be super useful.

Reply:
done.